tests/test_arithmetics.rs

```diff
@@ -186,6 +188,26 @@ fn binary_arithmetic() {
     py_run!(py, c, "assert 1 | c == '1 | BA'");
 }
 
+#[pyclass]
+struct RhsArithmetic {}
+
+#[pyproto]
+impl PyNumberProtocol for RhsArithmetic {
+    fn __radd__(&self, other: &PyAny) -> PyResult<String> {
+        Ok(format!("{:?} + RA", other))
+    }
+}
+
+#[test]
+fn rhs_arithmetic() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let c = Py::new(py, RhsArithmetic {}).unwrap();
+    py_run!(py, c, "assert c.__radd__(1) == '1 + RA'");
+    py_run!(py, c, "assert 1 + c == '1 + RA'");
```

Comment:
I think this patch isn't quite complete... The test case on line 207 passes while line 208 fails with this error:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'RhsArithmetic'
```

For some reason `1 + c` is not picking up the `__radd__` implementation, even though calling it directly works. I can do some more work on this if someone could give me some pointers on where to look next.

Reply:
If I understand the cpython code correctly, there's no `__radd__` in the c api and pyo3's way of implementing `__radd__` is actually fundamentally broken: It would need to share the `tp_as_number->nb_add` slot with `__add__`, with pyo3 dispatching to either `__add__` or `__radd__` depending on whether self or other is the first argument.

Since no one reported this in the two years that `__radd__` is part of pyo3, I think the best is just to comment out that line, merge this pull request and open an issue about reflected operations being broken.

Reply:
Now that I think about it, the way currently `__add__` works in `pyo3` is kind of magical: it allows `LHS` and `RHS` to be arbitrary types that may not even be the actual `Self` type of `PyNumberProtocol`'s impl. So I imagine `pyo3` must be doing some sort of smart dispatching to `__add__` already based on the lhs and rhs type information.

Comparing this to a plain python `__add__` implementation, if I understand things correctly, the python interpreter always passes lhs as `self` to `__add__` first, unless `__add__` returns `NotImplemented`, then it tries `__radd__` with the rhs as `self` to that method.